File: most_likely.py
import functools

@functools.lru_cache()
def lev(a, b):
    len_a = len(a)
    len_b = len(b)
    
    if len_b == 0: return len_a
    if len_a == 0: return len_b
    
    if a[0] == b[0]:
        return lev(a[1:], b[1:])
    
    return 1 + min([ lev(a[1:], b), lev(a, b[1:]), lev(a[1:], b[1:]) ])

commands = {
    'push': 0,
    'pull': 0,
    'commit': 0,
    'update': 0,
    'new': 0,
    'init': 0
}

text = input()
    
commands = { c: lev(c, text) for c in commands.keys() }
    
val = min(commands, key=commands.get)

if commands[val] == 0:
    print(val)
else:
    print('Incorrect command, did you mean "%s"?' % val)

File: MostLikely.hs
lev :: [Char] -> [Char] -> Int
lev a b
    | length b == 0    = length a
    | length a == 0    = length b
    | head a == head b = lev ta tb
    | otherwise        = 1 + foldr1 min [lev ta b, lev a tb, lev ta tb]
    where
        ta = tail a
        tb = tail b

commands = ["push", "pull", "commit", "update", "fetch", "init"]

main :: IO ()
main = do
    input <- getLine
    let computed = [(c, lev c input) | c <- commands]
    let closest = head (filter ((==) minsnd . snd) computed) where minsnd = minimum (map snd computed)
    
    putStrLn
        (case snd closest of
            0 -> fst closest
            _ -> "Incorrect command, did you mean '" ++ fst closest ++ "'?")
